File: tmf2/org.lttng.scope.tmf2.views.core/src/org/lttng/scope/tmf2/views/core/context/ViewGroupContext.java
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.tmf2.views.core.context;

import org.eclipse.jdt.annotation.Nullable;

import com.efficios.jabberwocky.common.TimeRange;
import com.efficios.jabberwocky.project.ITraceProject;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.property.SimpleObjectProperty;

/**
 * A common context for a group of views. Information is stored as properties,
 * and views can add listeners to get notified of value changes.
 *
 * @author Alexandre Montplaisir
 */
public class ViewGroupContext {

    /** Value representing uninitialized timestamps */
    public static final TimeRange UNINITIALIZED_RANGE = TimeRange.of(0, 0);

    private static final long DEFAULT_INITIAL_OFFSET = (1L * 100 * 1000 * 1000); // 100 ms (0.1 s)

    private final SignalBridge fSignalBridge;

    private final ObjectProperty<@Nullable ITraceProject<?, ?>> fCurrentTraceProject = new SimpleObjectProperty<>(null);

    private ObjectProperty<TimeRange> fCurrentVisibleTimeRange = new SimpleObjectProperty<>(UNINITIALIZED_RANGE);
    private ObjectProperty<TimeRange> fCurrentSelectionRange = new SimpleObjectProperty<>(UNINITIALIZED_RANGE);

    /**
     * The context is a singleton for now, but the framework could be extended
     * to support several contexts (one per "view group") at the same time.
     */
    private static @Nullable ViewGroupContext INSTANCE;

    private ViewGroupContext() {
        fSignalBridge = new SignalBridge(this);
    }

    /**
     * For now, there is only a single view context for the framework. This
     * method returns this singleton instance.
     *
     * @return The view context
     */
    public static ViewGroupContext getCurrent() {
        ViewGroupContext ctx = INSTANCE;
        if (ctx != null) {
            return ctx;
        }
        ctx = new ViewGroupContext();
        INSTANCE = ctx;
        return ctx;
    }

    /**
     * Cleanup all view group contexts.
     */
    public static void cleanup() {
        ViewGroupContext ctx = INSTANCE;
        if (ctx != null) {
            ctx.fSignalBridge.dispose();
        }
    }

    /**
     * Set the current trace being displayed by this view context.
     *
     * @param traceProject
     *            The trace project, can be null to indicate no trace
     */
    public void setCurrentTraceProject(@Nullable ITraceProject<?, ?> traceProject) {
        /* On trace change, adjust the other properties accordingly. */
        if (traceProject == null) {
            fCurrentVisibleTimeRange = new SimpleObjectProperty<>(UNINITIALIZED_RANGE);
            fCurrentSelectionRange = new SimpleObjectProperty<>(UNINITIALIZED_RANGE);
        } else {
            long start = traceProject.getStartTime();
            long end = traceProject.getEndTime();
            long visibleRangeEnd = Math.min(start + DEFAULT_INITIAL_OFFSET, end);

            fCurrentVisibleTimeRange = new SimpleObjectProperty<>((TimeRange.of(start, visibleRangeEnd)));
            fCurrentSelectionRange = new SimpleObjectProperty<>((TimeRange.of(start, start)));
        }

        fCurrentTraceProject.set(traceProject);
    }

    /**
     * Retrieve the current trace displayed by this view context.
     *
     * @return The context's current trace. Can be null to indicate no trace.
     */
    public @Nullable ITraceProject<?, ?> getCurrentTraceProject() {
        return fCurrentTraceProject.get();
    }

    /**
     * The current trace property.
     *
     * Make sure you use {@link #setCurrentTraceProject} to modify the current trace.
     *
     * @return The current trace property.
     */
    public ReadOnlyObjectProperty<@Nullable ITraceProject<?, ?>> currentTraceProjectProperty() {
        return fCurrentTraceProject;
    }

    /**
     * Set the current visible time range of this view context.
     *
     * @param range
     *            The new visible time range
     */
    public void setCurrentVisibleTimeRange(TimeRange range) {
        fCurrentVisibleTimeRange.set(range);
    }

    /**
     * Retrieve the current visible time range of the view context.
     *
     * @return The current visible time range
     */
    public TimeRange getCurrentVisibleTimeRange() {
        return fCurrentVisibleTimeRange.get();
    }

    /**
     * The visible time range property. This indicates the time range bounded by
     * the views of this context.
     *
     * @return The visible time range property
     */
    public ObjectProperty<TimeRange> currentVisibleTimeRangeProperty() {
        return fCurrentVisibleTimeRange;
    }

    /**
     * Set the current time range selection.
     *
     * @param range
     *            The new selection range
     */
    public void setCurrentSelectionTimeRange(TimeRange range) {
        fCurrentSelectionRange.set(range);
    }

    /**
     * Retrieve the current time range selection
     *
     * @return The current selection range
     */
    public TimeRange getCurrentSelectionTimeRange() {
        return fCurrentSelectionRange.get();
    }

    /**
     * The property representing the current time range selection. This is the
     * "highlighted" part of the trace, which can be selected by the mouse or
     * other means, and on which action can act.
     *
     * @return The time range selection property
     */
    public ObjectProperty<TimeRange> currentSelectionTimeRangeProperty() {
        return fCurrentSelectionRange;
    }

    /**
     * Utility method to get the full range of the current active trace project.
     *
     * @return The full range of the current trace project, or
     *         {@link #UNINITIALIZED_RANGE} if there is no active project.
     */
    public TimeRange getCurrentProjectFullRange() {
        ITraceProject<?, ?> project = getCurrentTraceProject();
        if (project == null) {
            return UNINITIALIZED_RANGE;
        }
        return TimeRange.of(project.getStartTime(), project.getEndTime());
    }
}

File: jabberwocky-core/src/main/kotlin/com/efficios/jabberwocky/context/ViewGroupContext.kt
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.efficios.jabberwocky.context;

import com.efficios.jabberwocky.common.TimeRange;
import com.efficios.jabberwocky.project.TraceProject;
import com.efficios.jabberwocky.trace.Trace
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyObjectProperty
import javafx.beans.property.SimpleObjectProperty;

/**
 * A common context for a group of views. Information is stored as properties,
 * and views can add listeners to get notified of value changes.
 *
 * Architecturally, the context of a given component should be constant (final)
 * for the lifetime of the component. If you wish to use a new context, then new
 * component instances should be created.
 *
 * The current project is not implemented using a standard JavaFX property, but instead
 * by implementing a [ProjectChangeListener] and registering it via [registerProjectChangeListener].
 *
 * If you attach change listeners to the time range properties, it is *very* important
 * to check the value of the [listenerFreeze] property within the listener. If it is true,
 * no action should be taken.
 * This is because some underlying operation - like a project-switching - is ongoing, and
 * the new values may not make sense on the current/previous project.
 */
class ViewGroupContext {

    companion object {
        /** Value representing uninitialized timestamps */
        @JvmField
        val UNINITIALIZED_RANGE = TimeRange.of(0, 0);
    }

    /**
     * Switching project is slightly more involved than just changing the value of a ObjectProperty.
     * Implementing this interface (including the 'flush()' method) will guarantee that only valid
     * values will ever be seen by listening components.
     */
    abstract class ProjectChangeListener(val owner: Any) {

        /**
         * Flush all currently queued operations that might work with the current project.
         *
         * For example, if queries are done in separate threads, implementation of the flush()
         * method should stop all currently queued threads.
         *
         * If a component does not use separate threads then this can be a no-op.
         */
        open fun flush() {
            /* Default implementation is to do nothing. Sub-classes can override. */
        }

        /**
         * Action to take on a project switching operation.
         *
         * The values of currentSelection and currentVisibleRange will be valid
         * once this is called.
         */
        abstract fun newProjectCb(newProject: TraceProject<*, *>?)
    }

    private val projectChangeListeners = mutableSetOf<ProjectChangeListener>()

    @Synchronized
    fun registerProjectChangeListener(listener: ProjectChangeListener): TraceProject<*, *>? {
        projectChangeListeners.add(listener)
        return traceProject
    }

    @Synchronized
    fun deregisterProjectChangeListener(listener: ProjectChangeListener) {
        projectChangeListeners.remove(listener)
    }

    @Volatile
    var listenerFreeze = false
        private set

    var traceProject: TraceProject<*, *>? = null
        private set

    @Synchronized
    fun switchProject(newProject: TraceProject<*, *>?) {
        val prevProject = traceProject
        if (prevProject != null && newProject != null) {
            /* Don't switch directly from one "real" project to another, go through a null project in-between. */
            switchProjectImpl(null)
            switchProjectImpl(newProject)
        } else {
            /* If we're going to or from a null-project, we can do it directly. */
            switchProjectImpl(newProject)
        }
    }

    private fun switchProjectImpl(newProject: TraceProject<*, *>?) {
        listenerFreeze = true
        projectChangeListeners.forEach { it.flush() }

        traceProject = newProject

        if (newProject == null) {
            visibleTimeRange = UNINITIALIZED_RANGE
            selectionTimeRange = UNINITIALIZED_RANGE
        } else {
            /* Initial visible range should be the whole project. Initial selection should be the first timestamp. */
            val start = newProject.startTime
            val end = newProject.endTime
            visibleTimeRange = TimeRange.of(start, end)
            selectionTimeRange = TimeRange.of(start, start)
        }

        projectChangeListeners.forEach { it.newProjectCb(newProject) }
        listenerFreeze = false
    }

    /**
     * Current visible time range in the context
     *
     * Do not forget to check the value of [listenerFreeze] inside any ChangeListener attached
     * to this property. If it is true, then no action should be taken using the value, because
     * it might not be consistent with the current project.
     * Once [listenerFreeze] becomes false again, then it is safe to use the values.
     */
    private val visibleTimeRangeProperty: ObjectProperty<TimeRange> = SimpleObjectProperty(UNINITIALIZED_RANGE)
    fun visibleTimeRangeProperty() = visibleTimeRangeProperty
    var visibleTimeRange: TimeRange
        get() = visibleTimeRangeProperty.get()
        set(value) = visibleTimeRangeProperty.set(value)

    /**
     * Current time range selection
     *
     * Do not forget to check the value of [listenerFreeze] inside any ChangeListener attached
     * to this property. If it is true, then no action should be taken using the value, because
     * it might not be consistent with the current project.
     * Once [listenerFreeze] becomes false again, then it is safe to use the values.
     */
    private val selectionTimeRangeProperty: ObjectProperty<TimeRange> = SimpleObjectProperty(UNINITIALIZED_RANGE)
    fun selectionTimeRangeProperty() = selectionTimeRangeProperty
    var selectionTimeRange: TimeRange
        get() = selectionTimeRangeProperty.get()
        set(value) = selectionTimeRangeProperty.set(value)


    /**
     * Utility method to get the full range of the current active trace project.
     *
     * @return The full range of the current trace project, or
     *         {@link #UNINITIALIZED_RANGE} if there is no active project.
     */
    fun getCurrentProjectFullRange(): TimeRange {
        return traceProject?.fullRange ?: UNINITIALIZED_RANGE
    }

    /**
     * Move the visible range to be centered on a target timestamp.
     *
     * The visible range span (or duration) will remain the same, we will simply "slide" the window.
     * This means it might be clamped to the end or start time of the project.
     *
     * Note this will not apply a selection on the target timetamp. If this is desired, it should
     * be done separately.
     */
    fun centerVisibleRangeOn(timestamp: Long) {
        val project = traceProject ?: return
        if (timestamp !in project.startTime..project.endTime) throw IllegalArgumentException("Target timestamp outside of project range")

        /* We will try our best to keep the current visible range span the same. */
        val span = visibleTimeRange.duration
        val halfSpan = span / 2
        visibleTimeRange = when {
            /* Clamp to start time */
            timestamp - halfSpan < project.startTime -> TimeRange.of(project.startTime, project.startTime + span)
            /* Clamp to end time */
            timestamp + halfSpan > project.endTime -> TimeRange.of(project.endTime - span, project.endTime)
            /* Simply center on the target timestamp, it should fit. */
            else -> TimeRange.of(timestamp - halfSpan, timestamp + halfSpan)
        }
    }
}
